controllers/nstemplateset/nstemplateset_controller_test.go

```diff
@@ -968,6 +968,63 @@ func TestDeleteNSTemplateSet(t *testing.T) {
 		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
 			DoesNotHaveFinalizer() // finalizer was not added and nothing else was done
 	})
+
+	t.Run("NSTemplateSet not deleted until namespace is deleted", func(t *testing.T) {
+		// given an NSTemplateSet resource and 1 active user namespaces ("dev")
+		nsTmplSet := newNSTmplSet(namespaceName, username, "advanced", withNamespaces("abcde11", "dev"), withDeletionTs(), withClusterResources("abcde11"))
+		devNS := newNamespace("advanced", username, "dev", withTemplateRefUsingRevision("abcde11"))
+
+		r, fakeClient := prepareController(t, nsTmplSet, devNS)
+		req := newReconcileRequest(namespaceName, username)
+
+		// only add deletion timestamp, but not delete
+		fakeClient.MockDelete = func(ctx context.Context, obj client.Object, opts ...client.DeleteOption) error {
+			if obj, ok := obj.(*corev1.Namespace); ok {
+				deletionTs := metav1.Now()
+				obj.DeletionTimestamp = &deletionTs
+				if err:= r.Client.Update(context.TODO(), obj); err !=nil {
+					return err
+				}
+			}
+			return nil
+		}
+		// first reconcile, deletion is triggered
+		_, err := r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+		//then second reconcile to check if namespace has actually been deleted
+		_, err = r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+
+		// get the first namespace and check that it is not deleted
+		firstNSName := fmt.Sprintf("%s-dev", username)
+		AssertThatNamespace(t, firstNSName, r.Client).DoesExist()
+		// get the NSTemplateSet resource again, check it is not deleted and its status
+		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).
+			HasFinalizer().
+			HasConditions(UnableToTerminate("user namespace johnsmith-dev deletion was triggered but is not complete yet, something could be blocking ns deletion"))
+
+		// remove deletion timestamp, actual delete is now triggered
+		userNS := &corev1.Namespace{}
+		err = r.Client.Get(context.TODO(), types.NamespacedName{Name: firstNSName}, userNS)
+		require.NoError(t, err)
+		userNS.DeletionTimestamp = nil
+		err = r.Client.Update(context.TODO(), userNS)
+		require.NoError(t, err)
+		fakeClient.MockDelete = nil //now removing the mockDelete
+
+		//reconcile, deletion is triggered
+		_, err = r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+		// second reconcile triggered when ns is deleted
+		_, err = r.Reconcile(context.TODO(), req)
+		require.NoError(t, err)
+
+		// get the first namespace and check it IS deleted
+		AssertThatNamespace(t, firstNSName, r.Client).DoesNotExist()
+		// Check that nsTemplateSet is set to delete
+		AssertThatNSTemplateSet(t, namespaceName, username, r.Client).DoesNotHaveFinalizer().HasConditions(Terminating())
+
```

Comment:
But this is not all.
Now you need to repeat all these steps from above for the second namespace and in the end make sure that the NSTemplateSet is finally deleted.

Reply:
the test case only has one namespace, so the NStemplateset is deleted after the first ns itself.

Reply:
Then we need to add another namespace to the test to cover all the usecases. Unless we already have such a test then why duplicating it here?

Reply:
correction : there are two namespaces in the testcase. 
There are existing tests - https://github.com/codeready-toolchain/member-operator/blob/master/controllers/nstemplateset/nstemplateset_controller_test.go#L870-L971  which check the deletion of NSTemplateSet. The reasons this new test case was added:
1. The existing tests checked that ns had deletionTimeStamp and not actual deletion
2. To test the negative scenario when the NSTemplateSet should not be deleted when the ns is not deleted.